### ISSUE TITLE
rejected execution (queue capacity 1000)

### ISSUE THREAD

**Issue by alexanderjarvis** [crate/crate#1316]
I'm running a perhaps complicated geo within query and getting a lot of errors. There are two queries run within 2 seconds of each other as well (with a slightly different bounding box as the polygon).
```
select from_loc, to_loc from travels  where ts > 1414972800000  and ts <= 1415059200000  and method = 5  and distance > 5  and duration > 5  and deleted != true  and within(to_loc, 'POLYGON ((-13.5791015625 55.727110085045986, 11.4697265625 55.727110085045986, 11.4697265625 47.57652571374621, -13.5791015625 47.57652571374621, -13.5791015625 55.727110085045986))')  LIMIT 100000
```

```
Caused by: org.elasticsearch.transport.RemoteTransportException: [Karima Shapandar][inet[/127.0.0.1:4300]][crate_sql] Caused by: io.crate.action.sql.SQLActionException: rejected execution (queue capacity 1000) on io.crate.action.sql.query.TransportQueryShardAction$AsyncAction$1@71a36902 	at io.crate.action.sql.TransportBaseSQLAction.buildSQLActionException(TransportBaseSQLAction.java:412) ~[na:na] 	at io.crate.action.sql.TransportBaseSQLAction.doExecute(TransportBaseSQLAction.java:179) ~[na:na] 	at io.crate.action.sql.TransportBaseSQLAction.doExecute(TransportBaseSQLAction.java:69) ~[na:na] 	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:65) ~[io.crate.crate-client-0.44.2.jar:na] 	at io.crate.action.sql.TransportSQLAction$TransportHandler.messageReceived(TransportSQLAction.java:154) ~[na:na] 	at io.crate.action.sql.TransportSQLAction$TransportHandler.messageReceived(TransportSQLAction.java:142) ~[na:na] 	at org.elasticsearch.transport.netty.MessageChannelHandler.handleRequest(MessageChannelHandler.java:217) ~[io.crate.crate-client-0.44.2.jar:na] 	at org.elasticsearch.transport.netty.MessageChannelHandler.messageReceived(MessageChannelHandler.java:111) ~[io.crate.crate-client-0.44.2.jar:na] 	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.elasticsearch.common.netty.OpenChannelsHandler.handleUpstream(OpenChannelsHandler.java:96) ~[io.crate.crate-client-0.44.2.jar:na] 	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:318) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108) ~[io.netty.netty-3.9.3.Final.jar:na] 	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42) ~[io.netty.netty-3.9.3.Final.jar:na] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) ~[na:1.7.0_65] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) ~[na:1.7.0_65] 	at java.lang.Thread.run(Thread.java:745) ~[na:1.7.0_65]  
```

I tried updating to the latest version of Crate and also increased the heap from 1G to 2G and still the same problem.

Any help appreciated!

**Comment by mfussenegger** [comment]
We've now a logic in place which greatly reduces this problem is it will reduce the amount of concurrency and process the shards one after another if the threadpool is already at its limits.

**Comment by tamaro** [comment]
Hello, 
was this implemented in crate 0.48.4 ?

thank you.

**Comment by seut** [comment]
This feature was released with 0.49.0.
